### PR TITLE
Add bidding history to user show page in admin

### LINKED ISSUE #775: Bid history shows on user page
As an admin user, I want to easily see the bid history for a vendor, so that I can can have some visibility into how the vendor is engaging with the micro-purchase platform.

Scenario: I see a user's bid history on the user show page
Given I am logged in as an admin
And there is a vendor in the system
And there have been auctions
And the vendor has participated in the auctions
When I go to the show page for that vendor
Then I should see a section labeled "Bid History"
And in that section I should see a table
And there should be a column labeled <bid_history_column_name>
And in that column I should see <bid_history_value>

| Bid History Column Name | Bid History Value |
| :----- | :----- |
| Auction | URL to auction show page |
| Auction Start Date | Start date of auction |
| Auction End Date | End date of auction |
| Number of Bids | The number of bids that vendor made on the auction |
| Won Auction? | Boolean showing whether vendor won the auction |
| Successful Delivery? | A Boolean showing whether the vendor successfully delivered on that auction |

### PR DESCRIPTION
**Pull request by harrisj**
Fixes #775

Generally I have tried to amend existing objects/methods where possible to add this new functionality. There are a few things worth mentioning that should be singled out for review:

1. I needed a good wrapper for displaying user-specific information about an auction (such as the user's bid count for the auction). I created a new class `Admin::UserAuctionViewModel` in the view_models, but I'm not too thrilled at my name for it.
2. I added a specific step definition file `admin_views_user_steps` because I had a step definition that was specific to that one feature file and this is the convention I've proposed for Cucumber steps are only linked to a specific feature.
3. The AuctionQuery object had defined a `my_bids` query, but it wasn't used anywhere in the code that I could see. So I renamed it and used it here (as well as added a test). This would be useful if we bring back the my_bids endpoint too.

### CODE REVIEW

File: app/models/auction_query.rb [hunk 1]
```diff
@@ -98,7 +98,7 @@ def public_find(id)
       .find(id)
   end
 
-  def my_bids(user_id)
+  def user_has_bid(user_id)
```
Comment (jessieay) [inline app/models/auction_query.rb]: maybe `user_bids` would be a more apt name for this given that it returns a collection of bids for a user?

File: app/view_models/admin/user_show_view_model.rb [hunk 2]
```diff
@@ -48,4 +48,12 @@ def contracting_officer
       "No"
     end
   end
+
+  def bids?
+    user_auctions.count > 0
+  end
+
+  def user_auctions
+    AuctionQuery.new.user_has_bid(user.id).map { |auction| Admin::UserAuctionViewModel.new(auction, user) }
```
Comment (jessieay) [inline app/view_models/admin/user_show_view_model.rb]: nice!

File: app/views/admin/users/show.html.erb [hunk 3]
```diff
@@ -9,3 +9,31 @@
   <li><b>Small business</b>: <%= @view_model.small_business %></li>
   <li><b>Contracting officer</b>: <%= @view_model.contracting_officer %></li>
 </ul>
+
+<% if @view_model.bids? %>
```
Comment (jessieay) [inline app/views/admin/users/show.html.erb]: let's use a partial instead! (in view model)

Comment (jessieay) [inline app/views/admin/users/show.html.erb]: eg: https://github.com/18F/micropurchase/blob/develop/app/view_models/auctions_index_view_model.rb#L31

**Comment by jessieay** [comment]
code climate error seems fixable!

File: app/view_models/admin/user_auction_view_model.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  def title
+    auction.title
+  end
+
+  def id
+    auction.id
+  end
+
+  def start_date
+    DcTimePresenter.convert_and_format(auction.started_at, DATE_FORMAT, timezone_label: false)
```
Comment (jessieay) [inline app/view_models/admin/user_auction_view_model.rb]: why not just show this in ET?

File: app/presenters/dc_time_presenter.rb [hunk 5]
```diff
@@ -26,9 +26,11 @@ def convert
     time.in_time_zone(time_zone)
   end
 
-  def convert_and_format(format = FORMAT)
+  def convert_and_format(format = FORMAT, timezone_label: true)
```
Comment (jessieay) [inline app/presenters/dc_time_presenter.rb]: why would we not want to show timezone label? seems potentially confusing

File: app/view_models/admin/user_auction_view_model.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    DcTimePresenter.convert_and_format(auction.ended_at, DATE_FORMAT, timezone_label: false)
+  end
+
+  def user_bid_count
+    user_bids.count
+  end
+
+  def user_won_label
+    if auction_over?
+      if user_won?
```
Comment (jessieay) [inline app/view_models/admin/user_auction_view_model.rb]: looks like `user_won?` already checks for `auction_over?`

File: app/views/admin/users/show.html.erb [hunk 7]
```diff
@@ -9,3 +9,31 @@
...
+
+<table class="usa-table-borderless" id="bidding-history">
+  <thead>
+    <tr>
+      <th scope="col">Auction</th>
+      <th scope="col">Start Date</th>
+      <th scope="col">End Date</th>
+      <th scope="col">User Bids</th>
+      <th scope="col">User Won?</th>
+      <th scope="col">Accepted?</th>
```
Comment (jessieay) [inline app/views/admin/users/show.html.erb]: Looks like these labels are slightly different from the ones in the original issue -- cool with me if you think these are clearer, just wanted to point that out in case it was an oversight

File: app/view_models/admin/user_auction_view_model.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      else
+        'No'
+      end
+    else
+      NA_RESPONSE_STRING
+    end
+  end
+
+  def accepted_label
+    if user_won?
```
Comment (jessieay) [inline app/view_models/admin/user_auction_view_model.rb]: this has more duplication but might be easier to read:

```
if user_won? && auction_accepted?
  'Yes'
elsif user_won?
  'No'
else
  NA_RESPONSE_STRING
end
```

File: app/models/auction_query.rb (hunk 1 above)
Comment (harrisj) [inline app/models/auction_query.rb]: It actually returns a collection of auctions where the user has placed a bid, so I struggled with the name a bit

File: app/views/admin/users/show.html.erb (hunk 3 above)
Comment (harrisj) [inline app/views/admin/users/show.html.erb]: Oh good point!

File: app/view_models/admin/user_auction_view_model.rb (hunk 4 above)
Comment (harrisj) [inline app/view_models/admin/user_auction_view_model.rb]: I am doing this so I will show it in ET (even though the started_at is stored in UTC)

File: app/presenters/dc_time_presenter.rb (hunk 5 above)
Comment (harrisj) [inline app/presenters/dc_time_presenter.rb]: I figured since this is an admin page it would be a little annoying to see 6/1/2016 EST and such over and over again

File: app/views/admin/users/show.html.erb (hunk 7 above)
Comment (harrisj) [inline app/views/admin/users/show.html.erb]: You are correct. I think this is clearer, but I'm willing to hear it out

File: app/views/admin/users/show.html.erb [hunk 9]
```diff
@@ -9,3 +9,5 @@
   <li><b>Small business</b>: <%= @view_model.small_business %></li>
   <li><b>Contracting officer</b>: <%= @view_model.contracting_officer %></li>
 </ul>
+
+<%= render partial: @view_model.bids_partial %>
```
Comment (jessieay) [inline app/views/admin/users/show.html.erb]: nice! I think you'll need to pass locals here too, right?

Comment (jessieay) [inline app/views/admin/users/show.html.erb]: oh wait I see it uses `@view_model` - this is super minor but I usually try to avoid instance vars in view partials (only use vars that are passed) - the benefit of that being that you only need to go to where something is rendered to know where the ivar comes from.

File: app/models/auction_query.rb (hunk 1 above)
Comment (jessieay) [inline app/models/auction_query.rb]: yeah that is tricky....what about something like `with_bid_for_user(user_id)` too wordy?

File: app/presenters/dc_time_presenter.rb (hunk 5 above)
Comment (jessieay) [inline app/presenters/dc_time_presenter.rb]: I think time zone is pretty helpful in admin. otherwise not sure if it is UTC or ET

Comment (harrisj) [inline app/presenters/dc_time_presenter.rb]: Okat, I added it back
